Add new spec for `go` package URLs (#338)
package-url/purl-spec: PURL-TYPES.rst

      pkg:github/package-url/purl-spec@244fd47e07d1004
      pkg:github/package-url/purl-spec@244fd47e07d1004#everybody/loves/dogs

go
------
``go`` for Go modules:

> - The ``namespace`` field is empty and implies the go mod proxy.

Reviewer: Should be done now, see the new commit (sorry for that).

> - The ``name`` will be the full module path.

Reviewer 2: Exactly. this is what the whole https://github.com/package-url/purl-spec/issues/308 is about. 
Please don't repeat the mistakes from the past.

> - The ``name`` will be the full module path.

Reviewer 2: ```suggestion
- The ``name`` is the full module path. In case of an URL: protocol MUST be lowercased; host-part MUST be lowercased; path-part MUSTbe unmodified, as it is case-sensitive.
```

Reviewer 3: I don't think this is correct.
1. I don't think it's legal to include a protocol in the module path. Go makes some HTTPS requests to resolve a VCS URL to download the package from (usually this is delegated to the proxy).
2. The host part is also part of the case sensitive module path. It should not be lowercased. Uppercase characters are currently forbidden by Go for modules. I don't think it's worthwhile or really correct for the PURL spec to be specifying how to convert an invalid module path into a valid module path, I don't think it's worthwhile for the PURL spec to be specifying how to validate Go module paths, this doesn't cover all the restrictions, and this may cause problems if Go ever changes the restrictions for some reason.

Reviewer 3: As far as Go is concerned, it's _usually_ a host-part but it has additional restrictions and it is case sensitive: https://go.dev/ref/mod#go-mod-file-ident

Reviewer 2: I see. I will modify my change-suggestion accordingly. does it fit better, now?

> - The ``name`` will be the full module path.

Reviewer: Should be done now, see the new commit (sorry for that).

> - The ``subpath`` will represent the package path within a module.

Reviewer 2: ```suggestion
- The ``subpath`` is the unmodified package path within a module.
```

> - The ``version`` will be a valid go version or pseudoversion, or empty.

Reviewer 2: ```suggestion
- The ``version`` may be a valid go version or pseudoversion, omitted when empty.
```

Reviewer: Why `may` here?

> - Additional Build information for binaries can be included as ``qualifiers`` (i.e VCS info, go version info, GoArch/GoOS info etc)

Reviewer 2: exactlty. be specific in the spec, so we all are on the same page.

> - Additional Build information for binaries can be included as ``qualifiers`` (i.e VCS info, go version info, GoArch/GoOS info etc)

Reviewer: PTAL in the new commit (sorry for that).

- Examples::

      pkg:go/google.golang.org%2Fgenproto#googleapis/api/annotations
      pkg:go/github.com%2Fjmorion%2Fsqlx@v1.1.2#api

>       pkg:go/golang.org%2Fx%2Fvuln?goversion=1.23.2&vcs=git&vcs_modified=true#cmd/govulncheck

Reviewer 4: There is a likely problem with the use of subpath: there is no way to determine where the module ends and the package starts in the general case, is there?
For instance, in the path `google.golang.org/genproto/googleapis/api/annotations` how can I determine safely that `google.golang.org/genproto` is a module and that `googleapis/api/annotations` is a package inside this module?  I need either a go proxy lookup or a full filesystem to locate a go.mod/go.sum file, right?

Reviewer 3: There is a way, if the module's code is available to you, to determine from a package import where module path ends and the package path begins by making HTTP requests.

I think the use of the subpath here is good because it puts the burden of determining this on whatever generates the PURL, which is likely aware of Go and either has the module paths or is most likely to be able to find the module path from the full package path. Then if you want to use a tool that checks PURLs against a database of information about modules (eg vulnerabilities), the tool already has all the information it needs. Otherwise, either the tool would need to make external API calls to figure out the module path of the PURL or the database would need to have an entry for every package in the module.

Reviewer: Just to add to @matt-phylum's comment. If a tool is producing a PURL for a Go artifact, then it can use `go version`, [Debug.BuildInfo](https://pkg.go.dev/debug/buildinfo#BuildInfo), or [packages.Load](https://pkg.go.dev/golang.org/x/tools/go/packages#Load) to get information about the package and its corresponding module. The encoding proposed here then makes it clear what the modules and packages are.

Reviewer 5: I don't believe this is true for the general use case of PURLs. E.g. we do static analysis of binaries and while we can get information about linked packages, there's no indication of which part of the paths correspond to modules

Reviewer: Right, if you are looking at a Go symbol from the symbol table, you can get its package. You can get the module correctly by prefix-matching it with module information from `debug.Buildinfo` of the binary, unless there are several modules that are prefixes of the package. My inclination is that it should not affect what is proposed here. (Arguably, there should be a way to get module information for a symbol in the binary, just the way one can do it for the source analysis.)

>       pkg:go/golang.org%2Fx%2Fvuln?goversion=1.23.2&vcs=git&vcs_modified=true#cmd/govulncheck

Reviewer 4: Is the plan to include all the buildinfo structure as qualifiers?
If so, this would only apply in a built binary?

Reviewer: We will expand on this.

>       pkg:go/golang.org%2Fx%2Fvuln@v1.1.3?goversion=1.23.2#cmd/govulncheck

Reviewer 4: Are the Go module versions always to be prefixed with a `v`?

Reviewer 3: A pseudoversion is a special kind of version that also starts with a v: https://go.dev/doc/modules/version-numbers#pseudo-version-number

I think for Go _modules_, including when using the Go module system to refer to something that predates modules, the version always starts with a v. In which case, versions that don't start with v would only be used with older tools like [Dep](https://github.com/golang/dep)?

Reviewer: If a version exists, it should be a valid Go [module version](https://go.dev/doc/modules/version-numbers).  It should start with a `v`.

Note that hashes should not be permitted, they are not a valid Go version (resolution of hash commits in go tooling is a convenience feature).
